### ISSUE TITLE
How is it better than the model of unittest or testament?

### ISSUE THREAD

**Issue by greenfork** [disruptek/balls#31]
Hi, I'm coming from a webdev background and I don't see clear use cases for why I would want to use this library. Maybe it will be beneficial to show this in documentation and educate similar-minded people like myself. Let me go through the goals of this project to better clarify my point of view:

- _expect everything to work; any statement is a test_  - how is that different from any other framework?
- _show the test code as we run it, unless not_ :wink:  - okay
- _better stack traces and test source output_  - cool
- _less magical syntax and less output omission_  - cool
- _aim to run many cheap tests in fewer files_  - what's the benefit in comparison with unittest?
- _easily accumulate state between tests when desired_  - why would I want to do that?
- _when attached to a tty, errors are not fatal_  - okay
- _individual tests don't even have to compile_  - don't see a use case
- _easier test reordering_  - easier compared to what? if we don't accumulate state, we don't care about test reordering
- _built-in timing, mem stats_  - maybe show it in an example, how is it better than existing benchmarking?
- _smaller, more self-evident testing macro_  - okay
- _no options means, "what would disruptek do?"_  - okay

Moreover in webdev world tests with accumulated state are shunned upon for the following reasons:
- Harder to reproduce since the state of a failed part depends upon previous execution
- Can't be run in parallel
- Hard to add new tests because they can potentially break tests before/after

In conclusion I don't see any significant advantages compared to what this model of testing would lose.

I understand that if I don't like it, I don't use it. But I believe this is just my insufficient understanding and I would really love to understand it better.

**Comment by disruptek** [comment]
I need to test code across multiple languages and memory models, and for me, compilation is the limiting factor when you use lots of test files.  I'm not as worried about contaminating state because that it easy to manage in Nim and it is, in fact, beneficial for speed and ease of writing tests.

This framework was developed to help debug CPS, which tends to fail compilation.  unittests is great at hiding test output and halting at the first compilation failure, which I find unhelpful.  The idea that _every statement is a test_ contrasts with unittests, in which any failure of a statement fails the encompassing test.  In this framework, each statement is itself a test.

The examples do show timing and memory stats, but they are really just a small window on performance that I find useful when comparing test runs quickly or comparing past runs in CI.  I use criterion for any serious benchmarking, but it's expensive for this application.

It's not obvious, but we _will_ be able to run our tests in parallel, and much faster than a typically threaded test runner.  I added threading to **testutils**, which is another test tool you should check out, so I've done it both ways.  This is better for me.

It's still a fairly new application and we're sorting out the goals and trying different things.  The codegen is more expensive than I'd like, so that may well change.  The test runner has been getting most of the attention lately because, again, what's important to me is to change the logic for what tests I run, and where, in only one place.  This does that.

I won't feel bad if you don't use the software, but know that the atypical design decisions are intentional -- there's no point in writing yet another unittests or testutils.  You can use them, or not -- nothing prevents you from using it exactly as you do unittests.

**Comment by greenfork** [comment]
> I need to test code across multiple languages

Do you use it to test other languages than Nim? As a side note, usually adoption of a foreign language for testing looks like the original language is not sufficient, it's like running nim forum on ruby on rails.

> I'm not as worried about contaminating state because that is easy to manage in Nim

"Easy to manage" - you mean using `block` and scoping in general? Or there are mechanisms in balls testing framework to achieve that?

> The examples do show timing and memory stats

Totally missed that, right!

>and for me, compilation is the limiting factor when you use lots of test files

Is this IO-bound or compiler has to recreate some runtime with every file which is the expensive part?

>what's important to me is to change the logic for what tests I run, and where, in only one place. This does that.

I'm not sure I understand it, there's `skip` and "run by pattern in files or test names" functionality in both unittest and testament which seems to do exactly the thing you describe here.

So far the enlightening point is:
- _"This framework was developed to help debug CPS, which tends to fail compilation. The idea that every statement is a test contrasts with unittests"_ - so my idea is that this framework is for testing whether something compiles or not as much as whether it has a correct output. And maybe the main use case is macros-heavy code or compilers?

**Comment by disruptek** [comment]
Maybe you should look at the code; the test runner is already quite a bit more complex than "runs a bunch of programs".

Yes, I do want to test C++ and JS as well as C.

Yes, I do want to test macro-heavy code and compilation.

It's relatively expensive to compile software versus running it, especially when you feel obliged to create a new program for every test.

There will be even better scope handling in the future; see the issues section.

The main use-case is that I want to change some code in one place and have all my software (in many other places) get tested differently.  I can't really do that without both a new runner and a new library.

**Comment by greenfork** [comment]
Thanks a lot for the explanations, that is enlightening in some ways. Maybe it could also be a good source for documentation.

So far I tried to test whether my raylib wrapper will compile with all the examples translated from C, didn't get much luck at it, the error message was too far above so I couldn't scroll to it after completion. Will try it in some time, hopefully it will be fixed
